### PR TITLE
[no release notes] Unit tests for KVSPuncherStore

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
* Learn how the puncher store behaves when clocks go wonky - this came up when I investigated possible ideas to address #1787 
* Add unit tests for a class that lacks them
* Document perhaps unexpected edge cases for the puncher's behaviour (_especially since the comments in the Puncher interface are not exactly correct under scenarios with clock drift_).

**Implementation Description (bullets)**:
* Add `KeyValueServicePuncherStoreTest`.

**Concerns (what feedback would you like?)**:
* Nothing for this PR, though I'm a bit concerned that our KVSPuncherStore does _NOT_ technically satisfy what the PuncherStore interface claims to be the case.

**Where should we start reviewing?**: the one file

**Priority (whenever / two weeks / yesterday)**: whenever

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1880)

### CODE REVIEW

**Comment by gsheasby** [comment]
Adding unit tests will decrease coverage? Good one, Codecov.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+public class KeyValueServicePuncherStoreTest {
+    private static final long TIMESTAMP_1 = 10L;
+    private static final long TIMESTAMP_2 = 20L;
+    private static final long TIMESTAMP_3 = 30L;
+    private static final long TIMESTAMP_BETWEEN_1_AND_2 = 15L;
+
+    private static final long WALL_CLOCK_1 = 100L;
+    private static final long WALL_CLOCK_2 = 200L;
+    private static final long WALL_CLOCK_3 = 300L;
+    private static final long WALL_CLOCK_BETWEEN_1_AND_2 = 150L;
```
Comment (gsheasby) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java]: smartassed nit: could use `(WALL_CLOCK_1 + WALL_CLOCK_2) / 2` (and analogously for the other two "between" constants.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        assertThat(puncherStore.getMillisForTimestamp(TIMESTAMP_BETWEEN_1_AND_2))
+                .isEqualTo(WALL_CLOCK_1);
+        puncherStore.put(TIMESTAMP_BETWEEN_1_AND_2, WALL_CLOCK_BETWEEN_1_AND_2);
+        assertThat(puncherStore.getMillisForTimestamp(TIMESTAMP_BETWEEN_1_AND_2))
+                .isEqualTo(WALL_CLOCK_BETWEEN_1_AND_2);
+    }
+
+    @Test
+    public void returnsGreatestPunchedTimeBeforeTimestampEvenIfNotAssociatedWithGreatestEligibleTimestamp() {
+        puncherStore.put(TIMESTAMP_BETWEEN_1_AND_2, WALL_CLOCK_BETWEEN_2_AND_3);
```
Comment (gsheasby) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java]: these cases could be made more obvious by using a different puncher store with a map of something like:
10 -> 200
20 -> 100.

Agreed that it's edge-casey. It's not completely obvious what it _should_ do in this circumstance. What it does is probably correct.

Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java]: Yep, agree that having a separate map would be much more explicit. I'll change this to do that instead.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java (hunk 1 above)
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java]: Indeed, changed!
